File: apps/hash-ai-worker-py/app/typesystem/data_type.py
"""A data type schema as defined by the Block Protocol."""
from typing import Literal

from pydantic import BaseModel, Extra, Field

from ._schema import OntologyTypeSchema

__all__ = ["DataTypeSchema"]


class DataTypeReference(BaseModel):
    ref: str = Field(..., alias="$ref")


class DataTypeSchema(OntologyTypeSchema, extra=Extra.allow):
    """A data type schema as defined by the Block Protocol.

    see https://blockprotocol.org/types/modules/graph/0.3/schema/data-type
    """

    kind: Literal["dataType"]
    ty: str = Field(..., alias="type")

File: libs/@blockprotocol/type-system/typescript/src/native/data-type.ts
import type { DataType, VersionedUrl } from "@blockprotocol/type-system-rs";

export const DATA_TYPE_META_SCHEMA: DataType["$schema"] =
  "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type";

/**
 * Returns all the IDs of all types referenced in a given data type.
 *
 * @param {DataType} dataType
 */
export const getReferencedIdsFromDataType = (
  dataType: DataType,
): {
  inheritsFromDataTypes: VersionedUrl[];
} => {
  const inheritsFromDataTypes: VersionedUrl[] = [];

  for (const inheritedEntityType of dataType.allOf ?? []) {
    inheritsFromDataTypes.push(inheritedEntityType.$ref);
  }

  return {
    inheritsFromDataTypes: [...inheritsFromDataTypes],
  };
};
